### ISSUE TITLE
How to integrate a custom "One liner annotation" rule

### ISSUE THREAD

**Issue by Tuubz** [detekt/detekt#2772]
Hello, everyone.

I just started out with implementing custom detekt rules. Besides all the existing rules, there's not really any other source I have found to help me with implement the following rule:

**VALID**
```
abstract class Test {
    @Inject lateinit var text: String
}
```

**INVALID**
```
abstract class Test {
    @Inject
    lateinit var text: String
}
```

Basically, I want all my `@Inject` variables to be written on one line

So far I have tried playing around with `visitAnnotationEntry(annotationEntry: KtAnnotationEntry)`, but I haven't quite figured out how to resolve this.

Any help is appreciated 🙏

**Comment by arturbosch** [comment]
Hi @Tuubz ,

The base idea would be to visit `KtProperty` nodes and see if the val/var line is the same as `property.annotationEntries.first()` line.

I currently do not have an idea here, so please
post if you need further assistance :).

**Comment by cortinico** [comment]
You can probably take inspiration from `AnnotationRule` from KtLint:
https://github.com/pinterest/ktlint/blob/master/ktlint-ruleset-experimental/src/main/kotlin/com/pinterest/ktlint/ruleset/experimental/AnnotationRule.kt

They're doing something really similar. You case is just a bit more strict so you should be able to adapt it easily.

**Comment by Tuubz** [comment]
Thanks @arturbosch & @cortinico 

I ended up with this solution:

```
package th.co.the1.the1app.rules

import io.gitlab.arturbosch.detekt.api.*
import org.jetbrains.kotlin.psi.KtElement
import org.jetbrains.kotlin.psi.KtProperty

class InjectAnnotationOneLine(config: Config) : Rule(config) {

    override val issue = Issue(
        id = javaClass.simpleName,
        severity = Severity.CodeSmell,
        description = ISSUE_DESCRIPTION,
        debt = Debt.FIVE_MINS
    )

    override fun visitKtElement(element: KtElement) {
        super.visitKtElement(element)
        if (element is KtProperty) {
            val isInjectProperty = element.text.contains(INJECT_ANNOTATION)
            if (isInjectProperty) {
                val isOneLineAnnotation = !element.text.contains("\n")
                if (!isOneLineAnnotation) {
                    val codeSmell = CodeSmell(
                        issue = issue,
                        entity = Entity.from(element),
                        message = CODE_SMELL_MESSAGE
                    )
                    report(codeSmell)
                }
            }
        }
    }
}

private const val INJECT_ANNOTATION = "@Inject"

private const val ISSUE_DESCRIPTION = "Issue: Inject Annotation One Line"
private const val CODE_SMELL_MESSAGE = "Inject annotations should be written on one line"
```

Any feedback for improvement is welcome 🙏

**Comment by arturbosch** [comment]
> Thanks @arturbosch & @cortinico
> 
> I ended up with this solution:
> 
> ```
> import io.gitlab.arturbosch.detekt.api.*
> import org.jetbrains.kotlin.psi.KtElement
> import org.jetbrains.kotlin.psi.KtProperty
> 
> class InjectAnnotationOneLine(config: Config) : Rule(config) {
> 
>     override val issue = Issue(
>         id = javaClass.simpleName,
>         severity = Severity.CodeSmell,
>         description = ISSUE_DESCRIPTION,
>         debt = Debt.FIVE_MINS
>     )
> 
>     override fun visitKtElement(element: KtElement) {
>         super.visitKtElement(element)
>         if (element is KtProperty) {
>             val isInjectProperty = element.text.contains(INJECT_ANNOTATION)
>             if (isInjectProperty) {
>                 val isOneLineAnnotation = !element.text.contains("\n")
>                 if (!isOneLineAnnotation) {
>                     val codeSmell = CodeSmell(
>                         issue = issue,
>                         entity = Entity.from(element),
>                         message = CODE_SMELL_MESSAGE
>                     )
>                     report(codeSmell)
>                 }
>             }
>         }
>     }
> }
> 
> private const val INJECT_ANNOTATION = "@Inject"
> 
> private const val ISSUE_DESCRIPTION = "Issue: Inject Annotation One Line"
> private const val CODE_SMELL_MESSAGE = "Inject annotations should be written on one line"
> ```
> 
> Any feedback for improvement is welcome

Looks good, you could be more explicit (and reduce code) by using constructs like:

```kotlin
override fun visitProperty(property: KtProperty) {
        super.visitProperty(property)
        property.annotations.any { it.text.contains("@Inject") }
    }
```
